### ISSUE TITLE
Supporting test output in XML

### ISSUE THREAD

**Issue by skyman** [TricksterGuy/complx#102]
Hey Brandon,

We're automating the grading of LC-3 assignments, and part of that is being able to read the test results. Right now, the tests results are produced in human-readable string format that's hard to parse. We want to be able to produce Ant-JUnit formatted XML (schema: https://raw.githubusercontent.com/windyroad/JUnit-Schema/master/JUnit.xsd) - this is the de-facto test result format used by Ant, Gradle, etc. and read by Jenkins, Bamboo, etc. and we already have built the tools to parse and display this automatically on our LMS software.

Now I wanted to do this myself - since you're already parsing XML for the test configs, the tools are there: it can't be too hard to produce XML. But I noticed that the XML reading abilities that you currently use depend on the wxWidgets library - and we do not want to have to install this library (so that we can build and deploy liblc3 and lc3test in a container environment without having to install wxWidgets and the ton of GTK-related dependencies it has).

As a result, I wanted to mark this as an issue here - it's kind of urgent since we need to be able to support this when homework 6 is released next week. The solution I had in mind was using something like TinyXML++ (https://github.com/wxFormBuilder/ticpp) instead of the wxWidgets XML API so that we're able to completely free the lc3test and liblc3 targets from the wxWidgets dependency, but before attempting to do anything, I wanted to ask you what you think, and how much you can help with this - I'm willing to attempt to do this myself if you're willing to merge it when it's done.

Thanks!

**Comment by TricksterGuy** [comment]
As far as me doing this I am not enthusiastic about it at all, and would like to see the framework the LMS provides before any decision is made on how to best go about making lc3test more useful for autograding purposes.

On that note to be honest I have been thinking  to totally scrap the xml parsing stuff in lieu for a friendlier interface to both TAs and students, as to what I have not decided. The first version of that XML processing stuff was written in about an hour at an airport and was a mess then and now is very unwieldy.

Also liblc3 does not have any dependencies.  lc3test like you said is dependent on wxWidgets, but all of that functionality is shoved into XMLTestParser.cpp

**Comment by TricksterGuy** [comment]
If you want to go that route then let me remove the xml stuff from pylc3.  As pylc3 requires that xmlparser.cpp file and thus the dependency on wxWidgets to even compile.

acutally I should remove the xml stuff anyway since pylc3 is bindings for liiblc3.

**Comment by skyman** [comment]
The point of doing the XML output in this format is not for the LMS systems but to meet the typical expectations from a software testing framework - since we're able to export test results in this format from many languages/platforms, it would make the most sense for lc3test to be capable of exporting in this format too. For the LMS system we use (Canvas) we still write an adapter on top of this, but that's acceptable from our perspective (as a result, instead of writing separate wrappers for JUnit, Check, and lc3test, we end up with one adapter since they would now all export the same format).

I think, at least for now, we can definitely follow you guys' suggestion, remove the wxWidgets dependency from pylc3, and I'll write a tester in Python. However, that too would use your XML file format for now (since I want students to be able to test locally using lc3test without Python), I'd just parse the XML in Python and run the setup using pylc3. Brandon, how soon can you remove the wxWidgets dependency for pylc3?

Overall, however, I think it would be good to start considering better ways of setting up tests as you said, and I really think giving the lc3test executable the ability to output in the format I mentioned above is the best for general compatibility of the file with the tons of testing-related software out there.

**Comment by TricksterGuy** [comment]
There is rule to not to ask me for any big/high priority stuff until December. As December is the only time of year I can really work on this. I've already uploaded the packages and only expected a minimal amount of work during this time of year.  Please give me more of an advance notice next time so I can prioritize things you all request vs other stuff that's needed.  I can't guarantee that this will be ready in time for homework 6, I'll make a best effort though.  I do expect a favor in return for all this.

The list of things that's needed for this is the following

0) Remove wxWidgets dependency - Done
1) Refactor test datastructure
2) Wrap datastructure fields to attributes in python
3) XML parser in python

From there you will be able to just run the test via python and export the data in however format you wish.

As far as Junit XML output, I'm not doing that now, that would be nice to have though. So you'll get that eventually.  If you want it sooner then you'll have to code that up yourself.

**Comment by TricksterGuy** [comment]
I've taken a look at the amount of work that needs to be done here and yeah unfortunately I am have to push back on this.  You are giving me like 4 days to make (what is turning out to be) a substantial rewrite of a portion of the application.  As I said before this area of code was not be best set of design decisions made.

Either way rewriting XMLTestParser.cpp to not use wxWidgets or further fleshing out pylc3 and removing the XMLTestParser dependency are both going to be a lot of work. The former would be faster however whatever xml library to replace it with would have to be learned quickly, the trouble with the latter is all of the refactoring that's needed here due to the way it was originally designed.

Furthermore I would also prefer to make these changes myself because of how poorly written it is and the fact that this area of code is not fully unit tested.

Remember that I am not compensated at all for any work I do here and that I have a job, and other obligations and stuff to work on other than 2110 related stuff.

All that said leaving the wxWidgets dependency in for now isn't going to end the world, this is the first semester it is introduced so everyone should expect for it to be a little rough around the edges.  The Travis CI thing that was added some time ago takes about 1 and a half minutes to run and yes running all of the unit tests also requires installing wxWidgets to build them. I honestly don't see this as a deal breaker, or at least you to asking me to fix this in the next 3 days.

The code is what it is though, and I'm steadily improving it, but this is not a reasonable request for me to do without more of an advance notice.

**Comment by skyman** [comment]
That's alright, really - I do understand this is bit of a short notice for such a drastic change. I'm willing to do most of this on my own, here's the current roadmap that I have:

* For this week’s assignment, I’m going to go ahead and fork complx and hack together something that produces non-Gradescope-compatible JSON, which is really not optional since we need to run lc3test with 4 files and merge the results anyway. So a Python script will grab this output and do the merging into a Gradescope-compatible JSON. At this point, it will remain an unofficial fork which I will attempt to compile just for the Gradescope grader, and students will never see this version outside of the Gradescope environment. Warning: it will probably be very slow since wxWidgets will be a dependency.

* Next, (presumably this upcoming weekend) I’m going to attempt to remove the wxWidgets dependency in lc3test in the config reader portion and replace it with two new config readers - an XML reader that parses the same current XML format just using a different library, and a YAML reader that parses the more readable (but for now, same structure) YAML format. At that point, I’ll attempt the first PR into your mainstream repo. This version will run much faster on Gradescope.

* Later on, I’ll also add the Ant JUnit XML exporter, refine the JSON output, and perhaps add the ability to run multiple test / asm pairs at the same time, producing Gradescope-compatible output without an adapter. Not sure when I’ll get to this but when I do, this will be the second PR into the mainstream repo. This version will also be the best choice for zucchini to support since the JUnit XML format will be already supported.

As a result, we're not going to need to use pyLC3 for grading - or at least, that's not the plan for now.

I promise you I'm going to try to make this as clean as possible, and hopefully, when I make the PRs, you can merge them.

**Comment by TricksterGuy** [comment]
As a general rule of thumb in the future, once I upload packages for Ubuntu I'm done with modifications to the program for that semester.

Alright as I said I would greatly prefer that I make these changes myself, but ok.

I would be interested in seeing how slow your first version will be.  I still don't see that as much of an issue since a student "grader" will still be provided yes? 


As a note If you are going to submit this as a PR I will not accept it without the following

1) You do not break any existing unit tests I have in place (of course!)
2) Anything you add must be fully unit tested.  Do not add to my technical debt. 
3) Any dependencies you add must be conditionally compilable with a flag in the CMakeLists.txt. I'd rather not have too many new required dependencies be added, this makes it more of a pain for me to produce a package for it.


I will probably be able to refactor the existing lc3_test data-structures in a few days I would say I am 40% there with that.  I would like to have more complete test coverage of XMLTestParser before you go and try writing your own.  Having that in before you start with that task though is not a guarantee to be in.  I would recommend you keep the same interface as that would be able to plug and play with the current tests that exist today.

**Comment by skyman** [comment]
So I finished the first milestone I mentioned above - I now have a hacked-together version (https://github.com/skyman/complx) that can only output JSON instead of the typical text output. I needed this one asap for the current upcoming assignment. It's pretty much temporary, but it works.

The docker image build on the LMS platform takes ~5 mins, but that's just run once - student submissions are graded in less than 5 seconds. It seems to work extremely well.

Since I'm satisfied with the way this works right now and the wxWidgets dependency doesn't appear to be an issue, I'm going to shuffle my roadmap and focus on fixing the outputs (XML, JSON) and make them work properly through a command line option, then I'll create a PR for that (it will satisfy the requirements you listed above).

Once that's done, I can also work on the new test data structure if you don't have time.

**Comment by TricksterGuy** [comment]
Yeah see, premature optimization is the root of all evil...

Unfortunately my refactoring is going to trample over your JSON stuff a little.  The revised code should be much cleaner though.  I may send a pull request over there if time permits.

I'm looking at being done with this tonight.  Right at 50%

**Comment by TricksterGuy** [comment]
It is almost 3am, this is the latest I can stay up.

The main refactoring is done, I'm now dealing with the fallout.  It will be in tomorrow night.

**Comment by skyman** [comment]
Awesome! I'm going to put in some work to this after you're done, this time I promise it won't be a hacky fix like before, haha... I can PR it into my repo and continue working on it if you don't have time to PR into mine after you're done with this.

and hey, thank you Brandon! This is pretty last minute and I couldn't expect you to help more, it's all very well appreciated.

**Comment by TricksterGuy** [comment]
Hmm... I honestly think you are making this harder on yourself, and looking at your commits we will butt heads over whether I accept that code you have written.

When Austin showed me that zucchini thing you both did,  that inspired me to start work on those python bindings since parsing the output was hard and I could have changed it at any time.  

There's no reason to modify lc3_test to spit out data in whatever format you want where you can get access to those data-structures directly and then do whatever you want with it from python.

One path in the future  I could take here is to just remove the xml files entirely, flesh out the python bindings and have all the testing be done from python.  If I actually had that done  I would pitch the idea on piazza and see how many students bite on that. Maybe summer 2018.

Anyway the path of least resistance would be to just use pylc3 with xml files (gotta revert that commit first). You would only have to create n LC3TestSuite objects.  Load the xml files and run the tests.  You can then aggregate the results and spit out the output.

But then again I don't know how this Gradescope container works.

And by the way all this work comes at a cost.  You did use up two days of my time...

I would say I'm at 95% now.

**Comment by skyman** [comment]
I don't think there's going to be any issues about merging the code I write, I had less than two hours to read through the source code and come up with the commit you read, and at no point did I claim it's quality work.

Problem with _just_ using pylc3 for testing comes from two points: we can no longer distribute tests to students as easily as we do now, and without the XML files, they wouldn't be able to run tests through the complx UI like they can right now.

So obviously there's a best of both worlds solution: supporting everything we can that's not too much work, so for example, the ability to input similarly structured XML, JSON and YML files containing the tests, using any of the lc3test, complx UI and pylc3 interfaces, and being able to export results in formats including readable text, JUnit XML and the Gradescope JSON format. For pyLC3, there would naturally be the ability to set up tests suites and read results programmatically without needing to use any of the formats mentioned above - but it's good if they're still usable from pylc3 too if needed.

I don't know about the Python binding part but I'm willing to cleanly write code for the import and export formats I mentioned above.

As for our Gradescope container code, I do plan to do as you said, and once this is all ready, use the pylc3 binding to read config files from XML and just get the outputs directly from the test suites from within the script. But for now I'm happy with my current solution as it ended up being extremely thin and neat on the Gradescope / python side: https://gist.github.com/skyman/589c797d0b0ae456164929e8ee270a98

**Comment by TricksterGuy** [comment]
Hmm... I think my changes are now unrelated to what you are doing, so far I've only refactored the internal test data-structures.  You just want a new output format? Now I am not feeling too okay with for reasons below.  I should really reverse that first commit I did so you should be able to use the XML files from within pylc3 again.

There are things about your commits that I do not like, and would do in a much, much different way.  But we will cross that bridge when we get there.  I do have some comments.

There is no way to run directly run XML tests in complx anymore, only through lc3_test. I removed that recently since it was no longer in parity with lc3_test, with how the results were presented to students I think the biggest annoyance was the quality of the questions asked when things failed. 

The only ability students have now is the ability to pick a test from an XML file and set up the same environment in complx.  If students want to run all the tests then lc3_test should be used.  Complx should be mainly used as a debugging tool.

But the general idea for the future here is, instead of distributing XML files that describe a test you will just distribute python files that describe (and run possibly?) the tests. Now an issue for me to resolve is how to still enable use of the above functionality when all I am given now is just a python script.  I suppose I these python scripts would have to be formatted a special way to be useful, and then I can interface with them via the C++ code.  But as I said still in idea stage.  I will have to think about that for a few weeks.  It is an interesting problem.

And well I don't like having all of these input/output formats exist here.  One day you will graduate and maybe you'll not care about this anymore.  However, I'm the captain of this ship and in this for as long as the ship sails, I don't want to have a big maintenance burden supporting all of these little formats just because a new TA comes along and introduces something because it is the new hotness of today.  So I'm thinking of not allowing THAT many input/output things, due to the reason that **I have to support that**.  I'm thinking as far as output maybe only the human readable format I have today and the Junit XML output since that's pretty much the standard and works with a lot of tools out there.  

The reason why I am more leaning toward the python stuff now because you guys can write all of your little tools and be done with it, and I don't have to be involved. Other users can do whatever they want and I don't have to be involved. It should be relatively low amount of effort to produce something with it.  All that is needed is the data-structures, of course one would need to learn how they are formatted.

**Comment by skyman** [comment]
This is addressed in #111.

**Comment by TricksterGuy** [comment]
unittest-xml-reporting Hype?

**Comment by TricksterGuy** [comment]
python branch merged
